leopard.nim

```diff
@@ -0,0 +1,315 @@
+  # done here then defining LeopardError as `object of CatchableError` will
+  # cause a mystery crash at compile-time (symbol not found). Can workaround by
+  # defining as just `object`, but then when trying to work with LeopardResult
+  # errors in e.g. tests/test_leopard.nim the same mystery crash happens at
+  # compile-time. The problem may be related to use of importcpp in
+  # leopard/wrapper.nim, so it could be a compiler bug. By redefining
+  # LeopardResult in this module (and casting wrapper.LeopardResult values) the
+  # the problem is avoided.
+  LeopardResult* = enum
+    LeopardNotEnoughData    = -11.cint # Buffer counts are too low
+    LeopardNeedLessData     = -10.cint # Too much recovery data received
+    LeopardInconsistentSize =  -9.cint # Buffer sizes must all be the same multiple of 64 bytes
+    LeopardBadCode          =  -8.cint # Bad RS code
+    LeopardCallInitialize   = wrapper.LeopardCallInitialize
+    LeopardPlatform         = wrapper.LeopardPlatform
+    LeopardInvalidInput     = wrapper.LeopardInvalidInput
+    LeopardInvalidCounts    = wrapper.LeopardInvalidCounts
+    LeopardInvalidSize      = wrapper.LeopardInvalidSize
+    LeopardTooMuchData      = wrapper.LeopardTooMuchData
+    LeopardNeedMoreData     = wrapper.LeopardNeedMoreData
+    LeopardSuccess          = wrapper.LeopardSuccess
+
+  LeopardError* = object of CatchableError
+    code*: LeopardResult
+
+  ParityData* = Data
+
+  ReedSolomonCode* = tuple[codeword, data, parity: uint] # symbol counts
+
+# https://github.com/catid/leopard/issues/12
+# https://www.cs.cmu.edu/~guyb/realworld/reedsolomon/reed_solomon_codes.html
+#
+# RS(255,239)
+# ---------------------------------
+# codeword symbols = 255
+# data symbols     = 239
+# parity symbols   = 255 - 239 = 16
+
+func isValid*(code: ReedSolomonCode): bool =
+  not ((code.codeword - code.data != code.parity) or
+       (code.parity > code.data) or (code.codeword < MinSymbols + 1) or
+       (code.data < MinSymbols) or (code.parity < MinSymbols) or
+       (code.codeword > MaxTotalSymbols))
+
+proc RS*(codeword, data: Positive): ReedSolomonCode =
+  var
+    parity = codeword - data
+
+  if parity < 0: parity = 0
+  (codeword: codeword.uint, data: data.uint, parity: parity.uint)
+
+when (NimMajor, NimMinor, NimPatch) < (1, 4, 0):
+  const
+    header = "<stdlib.h>"
+
+  proc c_malloc(size: csize_t): pointer {.importc: "malloc", header: header.}
+  proc c_free(p: pointer) {.importc: "free", header: header.}
+
+proc SIMDSafeAllocate(size: int): pointer {.inline.}  =
+  var
+    data =
+      when (NimMajor, NimMinor, NimPatch) < (1, 4, 0):
+        c_malloc(LEO_ALIGN_BYTES + size.uint)
+      else:
+        allocShared(LEO_ALIGN_BYTES + size.uint)
+
+    doffset = cast[uint](data) mod LEO_ALIGN_BYTES
+
+  data = offset(data, (LEO_ALIGN_BYTES + doffset).int)
+
+  var
+    offsetPtr = cast[pointer](cast[uint](data) - 1)
+
+  moveMem(offsetPtr, addr doffset, sizeof(doffset))
+  data
```

Comment:
I recommend using this directly for aligned allocation (from Weave, hence the wv_  prefix), no switch depending on Nim version.

```Nim
when defined(windows):
  proc aligned_alloc_windows(size, alignment: csize_t): pointer {.sideeffect,importc:"_aligned_malloc", header:"<malloc.h>".}
    # Beware of the arg order!
  proc wv_freeAligned*[T](p: ptr T){.sideeffect,importc:"_aligned_free", header:"<malloc.h>".}
elif defined(osx):
  proc posix_memalign(mem: var pointer, alignment, size: csize_t){.sideeffect,importc, header:"<stdlib.h>".}
  proc aligned_alloc(alignment, size: csize_t): pointer {.inline.} =
    posix_memalign(result, alignment, size)
  proc wv_freeAligned*[T](p: ptr T){.inline.} =
    c_free(p)
else:
  proc aligned_alloc(alignment, size: csize_t): pointer {.sideeffect,importc, header:"<stdlib.h>".}
  proc wv_freeAligned*[T](p: ptr T){.inline.} =
    c_free(p)

proc wv_allocAligned*(T: typedesc, alignment: static Natural): ptr T {.inline.} =
  ## aligned_alloc requires allocating in multiple of the alignment.
  static:
    assert alignment.isPowerOfTwo()
  let # TODO - cannot use a const due to https://github.com/nim-lang/Nim/issues/12726
    size = sizeof(T)
    requiredMem = size.roundNextMultipleOf(alignment)

  when defined(windows):
    cast[ptr T](aligned_alloc_windows(csize_t requiredMem, csize_t alignment))
  else:
    cast[ptr T](aligned_alloc(csize_t alignment, csize_t requiredMem))
```

In particular with allocShared, the GC will not return memory to the OS in a deterministic manner.
Furthermore, to make it threadsafe, there is a global lock hidden in allocShared.

Reply:
see also @arnetheduck comment: https://discord.com/channels/613988663034118151/616299964242460682/953921481052655616

> I'd go for c_malloc and it's friend posix_memalign for anything that's used in C, simply not to mix the nim allocator and its stack scanner with manually managed memory - also, if you're wrapping a C++ library, the recommendation would be to write a wrapper in C first that exposes the C++ code as C functions, then import those C functions into nim - this will save you a lot of trouble and headaches in general

Reply:
Thanks!

Reply:
> It's not supported well in Mingw windows

Do you know of or could you provide a summary of what the problems are?

Reply:
> Do you know of or could you provide a summary of what the problems are?

Unfortunately I don't remember well. That was years ago when trying to deal with OpenMP in Arraymancer. Now the issue might have been due to Nim and it's tlsEmulation flag which is/was default on Windows.

Reply:
Ah, good tip re: tlsEmulation. I'll try turning that off and check whether that improves the performance on Windows.

Reply:
@mratsim I've adapted `alloc/freeAligned` from weave and the changes are now on this PR branch: a4aebe89c0d9682f6104e2bd70d6d3ea0bc83ae7.

I'm happy to report that those changes dramatically improved the memory consumption story. On my local mac, when running the *"decode succeeds"* encode+decode test in a loop 100k or a millions times, the mem consumption barely gets above 35 MB!